lib/Service/RetentionService.php

```diff
@@ -0,0 +1,358 @@
+
+				// FIXME send notification
+				$this->sendReminder($user, $lastActivity, $policyDays);
+			} catch (SkipUserException $e) {
+				$this->logger->debug($e->getMessage(), $e->getLogParameters());
+				continue;
+			}
+		}
+
+		return true;
+	}
+
+	/**
+	 * @param IUser $user
+	 * @param int $skipIfNewerThan
+	 * @param ?int $skipIfOlderThan
+	 * @return int Return the last activity timestamp
+	 * @throws SkipUserException When the user should be skipped
+	 */
+	protected function shouldPerformActionOnUser(IUser $user, int $skipIfNewerThan, ?int $skipIfOlderThan = null): int {
+		$discoveryTimestamp = $this->skipUserBasedOnDiscovery($user);
+		$lastWebLogin = $user->getLastLogin();
+		$authTokensLastActivity = $this->getAuthTokensLastActivity($user);
+
+		$lastAction = max($discoveryTimestamp, $lastWebLogin, $authTokensLastActivity);
+
+		if ($this->keepUsersWithoutLogin && $lastAction === 0) {
+			throw new SkipUserException(
+				'Skipping user that never logged in: {user}',
+				['user' => $user->getUID()]
+			);
+		}
+
+		if ($skipIfNewerThan < $lastAction) {
+			throw new SkipUserException(
+				'Skipping user because last action is newer: {user}',
+				['user' => $user->getUID()]
+			);
+		}
+
+		if ($skipIfOlderThan !== null && $skipIfOlderThan > $lastAction) {
+			throw new SkipUserException(
+				'Skipping user because last action is older: {user}',
+				['user' => $user->getUID()]
+			);
+		}
+
+		return $lastAction;
+	}
+
+	/**
+	 * @param IUser $user
+	 * @return int Return the discovery timestamp as last activity timestamp
+	 * @throws SkipUserException When the user was just discovered
+	 */
+	protected function skipUserBasedOnDiscovery(IUser $user): int {
+		$discoveryTimestamp = (int) $this->config->getUserValue($user->getUID(), 'user_retention', 'user_created_at', '0');
+		if ($discoveryTimestamp === 0) {
+			// Set "now" as created at timestamp for the user.
+			$this->config->setUserValue($user->getUID(), 'user_retention', 'user_created_at', (string) $this->time->getTime());
+
+			throw new SkipUserException(
+				'New user, saving discovery time: {user}',
+				['user' => $user->getUID()]
+			);
+		}
+
+		return $discoveryTimestamp;
+	}
+
+	/**
+	 * @param IUser $user
+	 * @throws SkipUserException When the user is part of a group and should therefor be skipped
+	 */
+	protected function skipUserBasedOnProtectedGroupMembership(IUser $user): void {
+		if (empty($this->excludedGroups)) {
+			return;
+		}
+
+		$userGroups = $this->groupManager->getUserGroupIds($user);
+		$excludedGroups = array_intersect($userGroups, $this->excludedGroups);
+		if (!empty($excludedGroups)) {
+			throw new SkipUserException(
+				'Skipping user because of excluded groups ({groups}): {user}',
+				[
+					'user' => $user->getUID(),
+					'groups' => implode(',', $excludedGroups),
+				]
+			);
+		}
+	}
+
+	protected function getAuthTokensLastActivity(IUser $user): ?int {
```

Comment:
```suggestion
	protected function getAuthTokensLastActivity(IUser $user): int {
```
Either this returns 0 instead of null or you need to set default value when calling, but passing null to max() call seems dangerous.

Reply:
Ah, great catch.
It must return null, because 0 triggers another path (skip without login logic), but we should indeed filter it out before doing the max call